File: python/day07/part1.py
#!/usr/bin/env python3

class Folder:
    def __init__(self, name, parent=None):
        self.name = name
        self.parent = parent
        self.files = []
        self.folders = []
        self.size = None


def parse(input_path):
    with open(input_path, 'r') as f:
        return [[row.split(' ') for row in command.strip().split('\n')] for command in f.read().split('$')[1:]]


def compute_folder_sizes(folder):
    if folder.size is None:
        folder.size = sum([f[0] for f in folder.files]) + sum([compute_folder_sizes(f) for f in folder.folders])
    return folder.size


def build_tree(commands):
    root = Folder('/')
    curr = root
    for command in commands:
        if command[0][0] == 'ls':
            if len(curr.folders) == 0 and len(curr.files) == 0:
                for result in command[1:]:
                    if result[0] == 'dir':
                        curr.folders.append(Folder(result[1], curr))
                    else:
                        curr.files.append((int(result[0]), result[1]))
        elif command[0][0] == 'cd':
            if command[0][1] == '/':
                curr = root
            elif command[0][1] == '..':
                curr = curr.parent
            else:
                for folder in curr.folders:
                    if folder.name == command[0][1]:
                        curr = folder
                        break
    compute_folder_sizes(root)
    return root


def get_sum(folder):
    return (folder.size if folder.size <= 100000 else 0) + sum([get_sum(f) for f in folder.folders])


def solve_part1(commands):
    return get_sum(build_tree(commands))


if __name__ == "__main__":
    parsed = parse("data.txt")
    print('Part 1 :', solve_part1(parsed))

File: c++/day07/part1.cpp
#include "part1.h"

using namespace std;


namespace Day07 {

    void compute_size(shared_ptr<Directory> directory) {
        if (directory->size == 0) {
            for (const auto& [name, size] : directory->files) {
                directory->size += size;
            }
            for (auto& [name, folder] : directory->folders) {
                compute_size(folder);
                directory->size += folder->size;
            }
        }
    }

    shared_ptr<Directory> parse(const string& filename) {
        const auto lines = getFileLines(filename);
        const shared_ptr<Directory> directory = make_shared<Directory>("/");
        shared_ptr<Directory> curr_dir { directory };

        int curr_line = 1;
        while (curr_line < lines.size()) {
            auto tokens = split(lines[curr_line], " ");
            if (tokens[1] == "cd") {
                curr_dir = tokens[2] == ".." ? curr_dir->parent.lock() : curr_dir->folders.at(tokens[2]);
                ++curr_line;
                continue;
            } 

            // ls
            ++curr_line;
            while (curr_line < lines.size() && *lines[curr_line].begin() != '$') {
                tokens = split(lines[curr_line], " ");
                if (tokens[0] == "dir" && curr_dir->folders.find(tokens[1]) == curr_dir->folders.end()) {
                    curr_dir->folders.insert({ tokens[1], make_shared<Directory>(tokens[1]) });
                    curr_dir->folders.at(tokens[1])->parent = curr_dir;
                } else if (curr_dir->files.find(tokens[1]) == curr_dir->files.end()) {
                    curr_dir->files.insert({ tokens[1], stoi(tokens[0]) });
                }
                ++curr_line;
            }
        }

        compute_size(directory);

        return directory;
    }


    namespace Part1 {

        void calculate(const shared_ptr<Directory>& directory, int& result) {
            if (directory->size <= 100'000) {
                result += directory->size;
            }
            for (const auto& [_, folder] : directory->folders) {
                calculate(folder, result);
            }
        }

        int solve(const shared_ptr<Directory>& directory) {
            int result = 0;
            calculate(directory, result);
            return result;
        }
    }
}
